{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import requests\n",
    "import json\n",
    "from pandas.io.json import json_normalize\n",
    "sys.path.insert(0,'/home/nuria/workspace/utils3/lib/')\n",
    "import abravo_lib as utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# api addresses\n",
    "biolink = 'https://api.monarchinitiative.org/api'\n",
    "\n",
    "# endpoint\n",
    "endpoint = '/association'\n",
    "\n",
    "# path to data\n",
    "path = os.getcwd() + '/find_associations'\n",
    "if not os.path.isdir(path): os.makedirs(path)\n",
    "\n",
    "# read data\n",
    "sys.path.insert(0, './find_associations/')\n",
    "#/workspace/repurposing/rephetio-su/ngly1-net/networks/curated_1shellMonarchExpansion/data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# read ngly1 network nodes list\n",
    "# ngly1 network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  2.29it/s]\n"
     ]
    }
   ],
   "source": [
    "# get associations from Monarch: api\n",
    "nodes_l = ['NCBIGene:55768'] * 1\n",
    "\n",
    "# Monarch edges - compact version\n",
    "data = {'use_compact_associations':'true','fl_excludes_evidence':'true'}\n",
    "for node in tqdm(nodes_l, total = len(nodes_l)):\n",
    "    # out edges: from/\n",
    "    rc_out = requests.get('{}{}/from/{}'.format(biolink,endpoint,node),data)\n",
    "\n",
    "    # in edges: to/\n",
    "    rc_in = requests.get('{}{}/to/{}'.format(biolink,endpoint,node),data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'associations': None,\n",
       " 'compact_associations': [{'objects': ['GO:0046872', 'GO:0005515'],\n",
       "   'relation': 'RO:0002327',\n",
       "   'subject': 'NCBIGene:55768'},\n",
       "  {'objects': ['GO:0006457',\n",
       "    'KEGG-path:map04141',\n",
       "    'GO:0006517',\n",
       "    'REACT:R-HSA-532668'],\n",
       "   'relation': 'RO:0002331',\n",
       "   'subject': 'NCBIGene:55768'},\n",
       "  {'objects': ['OMIM:615273'], 'relation': None, 'subject': 'NCBIGene:55768'},\n",
       "  {'objects': ['GO:0005737', 'GO:0005634', 'GO:0005829'],\n",
       "   'relation': 'BFO:0000050',\n",
       "   'subject': 'NCBIGene:55768'}],\n",
       " 'facet_counts': {},\n",
       " 'facet_pivot': None,\n",
       " 'numFound': None,\n",
       " 'objects': None,\n",
       " 'start': None}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rc_out.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# save the result\n",
    "with open('{}/{}_out_compact_associations.json'.format(path, node), 'w') as f:\n",
    "    json.dump(rc_out.json(), f, sort_keys=True, indent=4)\n",
    "with open('{}/{}_in_compact_associations.json'.format(path, node), 'w') as f:\n",
    "    json.dump(rc_in.json(), f, sort_keys=True, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rc_out_df = json_normalize(rc_out.json(), 'compact_associations')\n",
    "rc_out_df.head(2)\n",
    "rc_out_df = rc_out_df.rename(\n",
    "                     columns= {\n",
    "                         'subject': 'node_queried_id',\n",
    "                         'relation':'relation_id',\n",
    "                         'objects': 'node_associated_id'\n",
    "                     }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rc_in_df = json_normalize(rc_in.json(), 'compact_associations')\n",
    "rc_in_df.head(2)\n",
    "rc_in_df = rc_in_df.rename(\n",
    "                     columns= {\n",
    "                         'objects': 'node_queried_id',\n",
    "                         'relation':'relation_id',\n",
    "                         'subject': 'node_associated_id'\n",
    "                     }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_queried_id</th>\n",
       "      <th>relation_id</th>\n",
       "      <th>node_associated_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>RO:0002327</td>\n",
       "      <td>[GO:0046872, GO:0005515]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>RO:0002331</td>\n",
       "      <td>[GO:0006457, KEGG-path:map04141, GO:0006517, R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>None</td>\n",
       "      <td>[OMIM:615273]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>BFO:0000050</td>\n",
       "      <td>[GO:0005737, GO:0005634, GO:0005829]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[NCBIGene:55768]</td>\n",
       "      <td>None</td>\n",
       "      <td>Coriell:GM25340</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_queried_id  relation_id  \\\n",
       "0    NCBIGene:55768   RO:0002327   \n",
       "1    NCBIGene:55768   RO:0002331   \n",
       "2    NCBIGene:55768         None   \n",
       "3    NCBIGene:55768  BFO:0000050   \n",
       "4  [NCBIGene:55768]         None   \n",
       "\n",
       "                                  node_associated_id  \n",
       "0                           [GO:0046872, GO:0005515]  \n",
       "1  [GO:0006457, KEGG-path:map04141, GO:0006517, R...  \n",
       "2                                      [OMIM:615273]  \n",
       "3               [GO:0005737, GO:0005634, GO:0005829]  \n",
       "4                                    Coriell:GM25340  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "associations_df = pd.concat([rc_out_df, rc_in_df], ignore_index=True)\n",
    "associations_df = associations_df[['node_queried_id', 'relation_id', 'node_associated_id']]\n",
    "associations_df.to_csv('{}/{}_compact_associations.tsv'.format(path, node), sep='\\t', index=False, header=True)\n",
    "associations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:01<00:00,  1.28s/it]\n"
     ]
    }
   ],
   "source": [
    "# Monarch edges - full version\n",
    "data = {'fl_excludes_evidence':'true'}\n",
    "for node in tqdm(nodes_l, total = len(nodes_l)):\n",
    "    # out edges: from/\n",
    "    r_out = requests.get('{}{}/from/{}'.format(biolink,endpoint,node),data)\n",
    "\n",
    "    # in edges: to/\n",
    "    r_in = requests.get('{}{}/to/{}'.format(biolink,endpoint,node),data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'associations': [{'evidence_graph': {'edges': [{'obj': 'ECO:0000304',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:431a40f1d20afbb9e4c94aabb82bd8f2a7053342'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:431a40f1d20afbb9e4c94aabb82bd8f2a7053342'},\n",
       "     {'obj': 'REACT:R-HSA-532668',\n",
       "      'pred': 'RO:0002331',\n",
       "      'sub': 'NCBIGene:55768'},\n",
       "     {'obj': 'REACT:R-HSA-532668',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:431a40f1d20afbb9e4c94aabb82bd8f2a7053342'}],\n",
       "    'nodes': [{'id': 'ECO:0000304',\n",
       "      'lbl': 'traceable author statement used in manual assertion'},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': 'MONARCH:431a40f1d20afbb9e4c94aabb82bd8f2a7053342', 'lbl': None},\n",
       "     {'id': 'REACT:R-HSA-532668',\n",
       "      'lbl': 'N-glycan trimming in the ER and Calnexin/Calreticulin cycle'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': '7b9b7ffd-9e85-494e-8f49-e5fbb7b7d508',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'REACT:R-HSA-532668',\n",
       "    'label': 'N-glycan trimming in the ER and Calnexin/Calreticulin cycle',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/reactome.ttl'],\n",
       "   'publications': None,\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'RO:0002331',\n",
       "    'label': 'involved in',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': ':.well-known/genid/genoGM25347',\n",
       "      'pred': 'GENO:0000222',\n",
       "      'sub': ':.well-known/genid/person-3200-1'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000418',\n",
       "      'sub': 'ClinVarVariant:126423'},\n",
       "     {'obj': ':.well-known/genid/610661.0004-610661.0005',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/genoGM25990'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'GENO:0000840',\n",
       "      'sub': 'ClinVarVariant:50962'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'RO:0002200',\n",
       "      'sub': ':.well-known/genid/person-3200-1'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'RO:0002200',\n",
       "      'sub': ':.well-known/genid/OMIM610661-615273VL'},\n",
       "     {'obj': ':.well-known/genid/person-3200-1',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:42d05e731935a651eaf999c2aeb046987f01cca7'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:a78544aa0feaec55807c289ba3d6aedadb69dbd8'},\n",
       "     {'obj': ':.well-known/genid/genoGM25340',\n",
       "      'pred': 'GENO:0000222',\n",
       "      'sub': ':.well-known/genid/person-3199-1'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'RO:0002200',\n",
       "      'sub': ':.well-known/genid/person-3199-1'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000408',\n",
       "      'sub': ':.well-known/genid/406885-404454VL'},\n",
       "     {'obj': ':.well-known/genid/406885-404454VL',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:665c1d3fbfc3abb0c5dc366dba0731358265d0f2'},\n",
       "     {'obj': ':.well-known/genid/OMIM610661-615273VL',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:980debcc2bf23bbd566978b66e42e8d7034e0df5'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:980debcc2bf23bbd566978b66e42e8d7034e0df5'},\n",
       "     {'obj': 'OMIM:610661.0005',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/610661.0004-610661.0005'},\n",
       "     {'obj': ':.well-known/genid/person-3199-1',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:4d8303fb4655006e8df2cb50e32740b91d6c3e65'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:42d05e731935a651eaf999c2aeb046987f01cca7'},\n",
       "     {'obj': 'OMIM:610661.0004',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/610661.0004-610661.0005'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'RO:0002200',\n",
       "      'sub': ':.well-known/genid/person-3213-1'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'RO:0002607',\n",
       "      'sub': ':.well-known/genid/KEGG-hsa55768-OMIM615273VL'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:8ece42278dc99592834ea5e667d42545d79f409a'},\n",
       "     {'obj': ':.well-known/genid/person-GM25990',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:a78544aa0feaec55807c289ba3d6aedadb69dbd8'},\n",
       "     {'obj': ':.well-known/genid/610661.0002',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/genoGM25347'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000408',\n",
       "      'sub': 'OMIM:610661.0001'},\n",
       "     {'obj': 'OMIM:610661.0002',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/610661.0001-610661.0002'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000408',\n",
       "      'sub': 'OMIM:610661.0002'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:665c1d3fbfc3abb0c5dc366dba0731358265d0f2'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000418',\n",
       "      'sub': ':.well-known/genid/OMIM610661-615273VL'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000408',\n",
       "      'sub': 'OMIM:610661.0003'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'GENO:0000840',\n",
       "      'sub': 'ClinVarVariant:126423'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'RO:0002200',\n",
       "      'sub': ':.well-known/genid/person-GM25990'},\n",
       "     {'obj': 'ECO:0000322',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:665c1d3fbfc3abb0c5dc366dba0731358265d0f2'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:8a568df7358124fda2f7fdd2cead58e43c25a34c'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'RO:0002200',\n",
       "      'sub': ':.well-known/genid/406885-404454VL'},\n",
       "     {'obj': 'OMIM:610661.0003',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/610661.0003'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:4d8303fb4655006e8df2cb50e32740b91d6c3e65'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'GENO:0000840',\n",
       "      'sub': 'ClinVarVariant:126424'},\n",
       "     {'obj': 'ECO:0000220',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:980debcc2bf23bbd566978b66e42e8d7034e0df5'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000418',\n",
       "      'sub': ':.well-known/genid/KEGG-hsa55768-OMIM615273VL'},\n",
       "     {'obj': ':.well-known/genid/genoGM25389',\n",
       "      'pred': 'GENO:0000222',\n",
       "      'sub': ':.well-known/genid/person-3213-1'},\n",
       "     {'obj': ':.well-known/genid/610661.0001-610661.0002',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/genoGM25340'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000408',\n",
       "      'sub': 'OMIM:610661.0005'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000408',\n",
       "      'sub': 'OMIM:610661.0004'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000418',\n",
       "      'sub': 'ClinVarVariant:126424'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000418',\n",
       "      'sub': 'ClinVarVariant:50962'},\n",
       "     {'obj': ':.well-known/genid/genoGM25990',\n",
       "      'pred': 'GENO:0000222',\n",
       "      'sub': ':.well-known/genid/person-GM25990'},\n",
       "     {'obj': ':.well-known/genid/person-3213-1',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:8ece42278dc99592834ea5e667d42545d79f409a'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000418',\n",
       "      'sub': 'ClinVarVariant:50961'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'GENO:0000840',\n",
       "      'sub': 'ClinVarVariant:126422'},\n",
       "     {'obj': ':.well-known/genid/KEGG-hsa55768-OMIM615273VL',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:8a568df7358124fda2f7fdd2cead58e43c25a34c'},\n",
       "     {'obj': 'OMIM:615273',\n",
       "      'pred': 'GENO:0000840',\n",
       "      'sub': 'ClinVarVariant:50961'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'GENO:0000418',\n",
       "      'sub': 'ClinVarVariant:126422'},\n",
       "     {'obj': 'OMIM:610661.0002',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/610661.0002'},\n",
       "     {'obj': ':.well-known/genid/610661.0003',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/genoGM25389'},\n",
       "     {'obj': 'OMIM:610661.0001',\n",
       "      'pred': 'GENO:0000382',\n",
       "      'sub': ':.well-known/genid/610661.0001-610661.0002'}],\n",
       "    'nodes': [{'id': ':.well-known/genid/person-3199-1',\n",
       "      'lbl': 'affected male proband with Congenital disorder of deglycosylation'},\n",
       "     {'id': 'ClinVarVariant:126423',\n",
       "      'lbl': 'NM_001145294.1(NGLY1):c.1079_1081delGAA (p.Arg360del)'},\n",
       "     {'id': ':.well-known/genid/person-3213-1',\n",
       "      'lbl': 'affected female proband with Congenital disorder of deglycosylation'},\n",
       "     {'id': ':.well-known/genid/610661.0004-610661.0005',\n",
       "      'lbl': 'NGLY1;NGLY1(3-BP DEL  1205 TTC;ARG542TER  c.1570C>T)'},\n",
       "     {'id': 'MONARCH:a78544aa0feaec55807c289ba3d6aedadb69dbd8', 'lbl': None},\n",
       "     {'id': 'ClinVarVariant:50961',\n",
       "      'lbl': 'NM_018297.3(NGLY1):c.1891delC (p.Gln631Serfs)'},\n",
       "     {'id': ':.well-known/genid/610661.0001-610661.0002',\n",
       "      'lbl': 'NGLY1;NGLY1(ARG401TER;c.1891delC)'},\n",
       "     {'id': ':.well-known/genid/person-GM25990',\n",
       "      'lbl': 'affected female proband with Congenital disorder of deglycosylation'},\n",
       "     {'id': 'OMIM:610661.0002', 'lbl': 'NGLY1, ARG401TER'},\n",
       "     {'id': ':.well-known/genid/610661.0002', 'lbl': 'NGLY1(ARG401TER)'},\n",
       "     {'id': 'ECO:0000322',\n",
       "      'lbl': 'imported manually asserted information used in automatic assertion'},\n",
       "     {'id': 'MONARCH:665c1d3fbfc3abb0c5dc366dba0731358265d0f2', 'lbl': None},\n",
       "     {'id': ':.well-known/genid/genoGM25347',\n",
       "      'lbl': 'NGLY1(ARG401TER) [GM25347]'},\n",
       "     {'id': 'MONARCH:8ece42278dc99592834ea5e667d42545d79f409a', 'lbl': None},\n",
       "     {'id': ':.well-known/genid/genoGM25340',\n",
       "      'lbl': 'NGLY1;NGLY1(ARG401TER;c.1891delC) [GM25340]'},\n",
       "     {'id': ':.well-known/genid/person-3200-1',\n",
       "      'lbl': 'affected male proband with Congenital disorder of deglycosylation'},\n",
       "     {'id': 'MONARCH:980debcc2bf23bbd566978b66e42e8d7034e0df5', 'lbl': None},\n",
       "     {'id': 'OMIM:610661.0003', 'lbl': 'NGLY1, 1-BP DUP, 1370G'},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': 'ClinVarVariant:126424',\n",
       "      'lbl': 'NM_001145293.1(NGLY1):c.1570C>T (p.Arg524Ter)'},\n",
       "     {'id': 'MONARCH:8a568df7358124fda2f7fdd2cead58e43c25a34c', 'lbl': None},\n",
       "     {'id': 'OMIM:615273', 'lbl': 'NGLY1-deficiency'},\n",
       "     {'id': ':.well-known/genid/KEGG-hsa55768-OMIM615273VL',\n",
       "      'lbl': 'some variant of K01456 peptide-N4-(N-acetyl-beta-glucosaminyl)asparagine amidase [EC:3.5.1.52] | (RefSeq) NGLY1'},\n",
       "     {'id': 'OMIM:610661.0001', 'lbl': 'NGLY1, 1-BP DEL, 1891C'},\n",
       "     {'id': ':.well-known/genid/genoGM25990',\n",
       "      'lbl': 'NGLY1;NGLY1(3-BP DEL  1205 TTC;ARG542TER  c.1570C>T) [GM25990]'},\n",
       "     {'id': 'ECO:0000220', 'lbl': 'sequencing assay evidence'},\n",
       "     {'id': ':.well-known/genid/610661.0003',\n",
       "      'lbl': 'NGLY1(c.1370dupG  R458FsTER)'},\n",
       "     {'id': 'MONARCH:4d8303fb4655006e8df2cb50e32740b91d6c3e65', 'lbl': None},\n",
       "     {'id': ':.well-known/genid/OMIM610661-615273VL',\n",
       "      'lbl': 'some variant of NGLY1 that causes Congenital disorder of deglycosylation'},\n",
       "     {'id': ':.well-known/genid/genoGM25389',\n",
       "      'lbl': 'NGLY1(c.1370dupG  R458FsTER) [GM25389]'},\n",
       "     {'id': ':.well-known/genid/406885-404454VL',\n",
       "      'lbl': 'some variant of NGLY1 that is a disease-causing germline mutation(s) (loss of function) in Alacrimia-choreoathetosis-liver dysfunction syndrome'},\n",
       "     {'id': 'ClinVarVariant:126422',\n",
       "      'lbl': 'NM_018297.3(NGLY1):c.1370dupG (p.Arg458Lysfs)'},\n",
       "     {'id': 'ClinVarVariant:50962',\n",
       "      'lbl': 'NM_018297.3(NGLY1):c.1201A>T (p.Arg401Ter)'},\n",
       "     {'id': 'OMIM:610661.0005', 'lbl': 'NGLY1, ARG542TER'},\n",
       "     {'id': 'MONARCH:42d05e731935a651eaf999c2aeb046987f01cca7', 'lbl': None},\n",
       "     {'id': 'OMIM:610661.0004', 'lbl': 'NGLY1, 3-BP DEL, 1205TTC'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': 'af89e914-68e5-4b3f-8d02-ee6d214a8a19',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'OMIM:615273',\n",
       "    'label': 'NGLY1-deficiency',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/kegg.ttl',\n",
       "    'http://data.monarchinitiative.org/ttl/orphanet.ttl',\n",
       "    'https://data.monarchinitiative.org/ttl/omim.ttl',\n",
       "    'http://data.monarchinitiative.org/ttl/coriell.ttl',\n",
       "    'http://data.monarchinitiative.org/ttl/clinvar.ttl'],\n",
       "   'publications': None,\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': None,\n",
       "    'label': None,\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': ':.well-known/genid/NCBIGene55768product',\n",
       "      'pred': 'RO:0002205',\n",
       "      'sub': 'NCBIGene:55768'},\n",
       "     {'obj': 'KEGG-path:map04141',\n",
       "      'pred': 'RO:0002331',\n",
       "      'sub': ':.well-known/genid/NCBIGene55768product'}],\n",
       "    'nodes': [{'id': 'KEGG-path:map04141',\n",
       "      'lbl': 'Protein processing in endoplasmic reticulum'},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': ':.well-known/genid/NCBIGene55768product', 'lbl': None}]},\n",
       "   'evidence_types': None,\n",
       "   'id': 'ab5c5681-ac35-4726-842e-cda2a1c72db7',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'KEGG-path:map04141',\n",
       "    'label': 'Protein processing in endoplasmic reticulum',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/ctd.ttl'],\n",
       "   'publications': None,\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'RO:0002331',\n",
       "    'label': 'involved in',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': 'ECO:0000304',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:f3fd59ce099c0e1c18def76ed26c245b9f385388'},\n",
       "     {'obj': 'ECO:0000318',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:f3fd59ce099c0e1c18def76ed26c245b9f385388'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:f3fd59ce099c0e1c18def76ed26c245b9f385388'},\n",
       "     {'obj': 'REACT:R-HSA-8850594',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:f3fd59ce099c0e1c18def76ed26c245b9f385388'},\n",
       "     {'obj': 'GO:0005829', 'pred': 'BFO:0000050', 'sub': 'NCBIGene:55768'},\n",
       "     {'obj': 'GO_REF:0000033',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:f3fd59ce099c0e1c18def76ed26c245b9f385388'},\n",
       "     {'obj': 'GO:0005829',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:f3fd59ce099c0e1c18def76ed26c245b9f385388'}],\n",
       "    'nodes': [{'id': 'ECO:0000304',\n",
       "      'lbl': 'traceable author statement used in manual assertion'},\n",
       "     {'id': 'MONARCH:f3fd59ce099c0e1c18def76ed26c245b9f385388', 'lbl': None},\n",
       "     {'id': 'REACT:R-HSA-8850594', 'lbl': None},\n",
       "     {'id': 'ECO:0000318',\n",
       "      'lbl': 'biological aspect of ancestor evidence used in manual assertion'},\n",
       "     {'id': 'GO_REF:0000033', 'lbl': None},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': 'GO:0005829', 'lbl': 'cytosol'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': '5dc9e10a-3e51-484f-8465-ad5c90507307',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'GO:0005829',\n",
       "    'label': 'cytosol',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['http://data.monarchinitiative.org/ttl/go.ttl'],\n",
       "   'publications': [{'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'REACT:R-HSA-8850594',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None},\n",
       "    {'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'GO_REF:0000033',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None}],\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'BFO:0000050',\n",
       "    'label': 'is part of',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': 'PMID:25416956',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:7e7b9207b54437658fd60deba03c5a99ecbd2593'},\n",
       "     {'obj': 'PMID:15358861',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:7e7b9207b54437658fd60deba03c5a99ecbd2593'},\n",
       "     {'obj': 'GO:0005515', 'pred': 'RO:0002327', 'sub': 'NCBIGene:55768'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:7e7b9207b54437658fd60deba03c5a99ecbd2593'},\n",
       "     {'obj': 'GO:0005515',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:7e7b9207b54437658fd60deba03c5a99ecbd2593'},\n",
       "     {'obj': 'ECO:0000353',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:7e7b9207b54437658fd60deba03c5a99ecbd2593'}],\n",
       "    'nodes': [{'id': 'MONARCH:7e7b9207b54437658fd60deba03c5a99ecbd2593',\n",
       "      'lbl': None},\n",
       "     {'id': 'PMID:25416956', 'lbl': None},\n",
       "     {'id': 'PMID:15358861', 'lbl': None},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': 'GO:0005515', 'lbl': 'protein binding'},\n",
       "     {'id': 'ECO:0000353',\n",
       "      'lbl': 'physical interaction evidence used in manual assertion'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': '8cc76b39-6347-47e4-a8cd-66d9fb32b32f',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'GO:0005515',\n",
       "    'label': 'protein binding',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['http://data.monarchinitiative.org/ttl/go.ttl'],\n",
       "   'publications': [{'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'PMID:25416956',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None},\n",
       "    {'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'PMID:15358861',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None}],\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'RO:0002327',\n",
       "    'label': 'enables',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': 'ECO:0000304',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:966450122b7930a0b28f4d5608c55bfaab35208e'},\n",
       "     {'obj': 'REACT:R-HSA-532668',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:966450122b7930a0b28f4d5608c55bfaab35208e'},\n",
       "     {'obj': 'GO:0006457', 'pred': 'RO:0002331', 'sub': 'NCBIGene:55768'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:966450122b7930a0b28f4d5608c55bfaab35208e'},\n",
       "     {'obj': 'GO:0006457',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:966450122b7930a0b28f4d5608c55bfaab35208e'}],\n",
       "    'nodes': [{'id': 'MONARCH:966450122b7930a0b28f4d5608c55bfaab35208e',\n",
       "      'lbl': None},\n",
       "     {'id': 'ECO:0000304',\n",
       "      'lbl': 'traceable author statement used in manual assertion'},\n",
       "     {'id': 'GO:0006457', 'lbl': 'protein folding'},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': 'REACT:R-HSA-532668',\n",
       "      'lbl': 'N-glycan trimming in the ER and Calnexin/Calreticulin cycle'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': 'd71829fb-2943-40a7-b19e-0ee97a4fc362',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'GO:0006457',\n",
       "    'label': 'protein folding',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['http://data.monarchinitiative.org/ttl/go.ttl'],\n",
       "   'publications': [{'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'REACT:R-HSA-532668',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None}],\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'RO:0002331',\n",
       "    'label': 'involved in',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:cf6192c9b876e8b73b8a5746308cc3c996590a14'},\n",
       "     {'obj': 'PMID:15358861',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:cf6192c9b876e8b73b8a5746308cc3c996590a14'},\n",
       "     {'obj': 'ECO:0000314',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:cf6192c9b876e8b73b8a5746308cc3c996590a14'},\n",
       "     {'obj': 'GO:0005737',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:cf6192c9b876e8b73b8a5746308cc3c996590a14'},\n",
       "     {'obj': 'GO:0005737', 'pred': 'BFO:0000050', 'sub': 'NCBIGene:55768'}],\n",
       "    'nodes': [{'id': 'PMID:15358861', 'lbl': None},\n",
       "     {'id': 'GO:0005737', 'lbl': 'cytoplasm'},\n",
       "     {'id': 'MONARCH:cf6192c9b876e8b73b8a5746308cc3c996590a14', 'lbl': None},\n",
       "     {'id': 'ECO:0000314',\n",
       "      'lbl': 'direct assay evidence used in manual assertion'},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': '5c91a2ae-43d4-414d-a5d1-b0b1387a5cd4',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'GO:0005737',\n",
       "    'label': 'cytoplasm',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['http://data.monarchinitiative.org/ttl/go.ttl'],\n",
       "   'publications': [{'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'PMID:15358861',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None}],\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'BFO:0000050',\n",
       "    'label': 'is part of',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:3d961412803b3aecb295c6f42517d292bae2c591'},\n",
       "     {'obj': 'GO:0046872',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:3d961412803b3aecb295c6f42517d292bae2c591'},\n",
       "     {'obj': 'ECO:0000501',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:3d961412803b3aecb295c6f42517d292bae2c591'},\n",
       "     {'obj': 'GO_REF:0000037',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:3d961412803b3aecb295c6f42517d292bae2c591'},\n",
       "     {'obj': 'GO:0046872', 'pred': 'RO:0002327', 'sub': 'NCBIGene:55768'}],\n",
       "    'nodes': [{'id': 'MONARCH:3d961412803b3aecb295c6f42517d292bae2c591',\n",
       "      'lbl': None},\n",
       "     {'id': 'ECO:0000501', 'lbl': 'evidence used in automatic assertion'},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': 'GO:0046872', 'lbl': 'metal ion binding'},\n",
       "     {'id': 'GO_REF:0000037', 'lbl': None}]},\n",
       "   'evidence_types': None,\n",
       "   'id': '68ed8bc7-b3cd-4377-8dbd-9c8bd8670e44',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'GO:0046872',\n",
       "    'label': 'metal ion binding',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['http://data.monarchinitiative.org/ttl/go.ttl'],\n",
       "   'publications': [{'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'GO_REF:0000037',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None}],\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'RO:0002327',\n",
       "    'label': 'enables',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': 'GO:0005634',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:8a7402d4856c77daa7951088992388590d070662'},\n",
       "     {'obj': 'GO_REF:0000033',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:8a7402d4856c77daa7951088992388590d070662'},\n",
       "     {'obj': 'ECO:0000318',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:8a7402d4856c77daa7951088992388590d070662'},\n",
       "     {'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:8a7402d4856c77daa7951088992388590d070662'},\n",
       "     {'obj': 'GO:0005634', 'pred': 'BFO:0000050', 'sub': 'NCBIGene:55768'}],\n",
       "    'nodes': [{'id': 'ECO:0000318',\n",
       "      'lbl': 'biological aspect of ancestor evidence used in manual assertion'},\n",
       "     {'id': 'GO_REF:0000033', 'lbl': None},\n",
       "     {'id': 'MONARCH:8a7402d4856c77daa7951088992388590d070662', 'lbl': None},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'},\n",
       "     {'id': 'GO:0005634', 'lbl': 'nucleus'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': '05c0834d-031e-46c7-96ed-679a264cfb0e',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'GO:0005634',\n",
       "    'label': 'nucleus',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['http://data.monarchinitiative.org/ttl/go.ttl'],\n",
       "   'publications': [{'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'GO_REF:0000033',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None}],\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'BFO:0000050',\n",
       "    'label': 'is part of',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None},\n",
       "  {'evidence_graph': {'edges': [{'obj': 'NCBIGene:55768',\n",
       "      'pred': 'OBAN:association_has_subject',\n",
       "      'sub': 'MONARCH:b88602ff81957469b90c4354cda719ece61671f9'},\n",
       "     {'obj': 'GO_REF:0000033',\n",
       "      'pred': 'dc:source',\n",
       "      'sub': 'MONARCH:b88602ff81957469b90c4354cda719ece61671f9'},\n",
       "     {'obj': 'GO:0006517',\n",
       "      'pred': 'OBAN:association_has_object',\n",
       "      'sub': 'MONARCH:b88602ff81957469b90c4354cda719ece61671f9'},\n",
       "     {'obj': 'ECO:0000318',\n",
       "      'pred': 'RO:0002558',\n",
       "      'sub': 'MONARCH:b88602ff81957469b90c4354cda719ece61671f9'},\n",
       "     {'obj': 'GO:0006517', 'pred': 'RO:0002331', 'sub': 'NCBIGene:55768'}],\n",
       "    'nodes': [{'id': 'MONARCH:b88602ff81957469b90c4354cda719ece61671f9',\n",
       "      'lbl': None},\n",
       "     {'id': 'ECO:0000318',\n",
       "      'lbl': 'biological aspect of ancestor evidence used in manual assertion'},\n",
       "     {'id': 'GO_REF:0000033', 'lbl': None},\n",
       "     {'id': 'GO:0006517', 'lbl': 'protein deglycosylation'},\n",
       "     {'id': 'NCBIGene:55768', 'lbl': 'NGLY1'}]},\n",
       "   'evidence_types': None,\n",
       "   'id': 'b449c1f2-570d-42cb-943c-dad3ca0f111f',\n",
       "   'object': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'GO:0006517',\n",
       "    'label': 'protein deglycosylation',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': None, 'label': None},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'object_extension': None,\n",
       "   'provided_by': ['http://data.monarchinitiative.org/ttl/go.ttl'],\n",
       "   'publications': [{'categories': None,\n",
       "     'consider': None,\n",
       "     'deprecated': None,\n",
       "     'description': None,\n",
       "     'id': 'GO_REF:0000033',\n",
       "     'label': None,\n",
       "     'replaced_by': None,\n",
       "     'synonyms': None,\n",
       "     'types': None}],\n",
       "   'qualifiers': None,\n",
       "   'relation': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'RO:0002331',\n",
       "    'label': 'involved in',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'types': None},\n",
       "   'slim': None,\n",
       "   'subject': {'categories': None,\n",
       "    'consider': None,\n",
       "    'deprecated': None,\n",
       "    'description': None,\n",
       "    'id': 'NCBIGene:55768',\n",
       "    'label': 'NGLY1',\n",
       "    'replaced_by': None,\n",
       "    'synonyms': None,\n",
       "    'taxon': {'id': 'NCBITaxon:9606', 'label': 'Homo sapiens'},\n",
       "    'types': None,\n",
       "    'xrefs': None},\n",
       "   'subject_extension': None,\n",
       "   'type': None}],\n",
       " 'compact_associations': None,\n",
       " 'facet_counts': {'object_closure': {'BFO:0000001': 73,\n",
       "   'BFO:0000002': 68,\n",
       "   'BFO:0000003': 31,\n",
       "   'BFO:0000004': 23,\n",
       "   'BFO:0000015': 10,\n",
       "   'BFO:0000020': 12,\n",
       "   'BFO:0000031': 41,\n",
       "   'BFO:0000040': 23,\n",
       "   'GENO:0000014': 31,\n",
       "   'GENO:0000701': 31,\n",
       "   'IAO:0000030': 10,\n",
       "   'OBO:CARO_0000006': 21,\n",
       "   'SO:0000001': 31,\n",
       "   'SO:0000110': 31,\n",
       "   'SO:0000704': 31,\n",
       "   'SO:0001217': 31,\n",
       "   'SO:0001411': 31,\n",
       "   'UBERON:0000061': 17,\n",
       "   'UBERON:0000064': 10,\n",
       "   'UBERON:0000465': 17,\n",
       "   'UBERON:0001062': 20,\n",
       "   'UBERON:0010000': 17,\n",
       "   'http://uri.neuinfo.org/nif/nifstd/birnlex_16': 10,\n",
       "   'http://uri.neuinfo.org/nif/nifstd/birnlex_6': 20,\n",
       "   'owl:Thing': 73},\n",
       "  'subject_taxon_label': {'Homo sapiens': 73}},\n",
       " 'facet_pivot': None,\n",
       " 'numFound': None,\n",
       " 'objects': None,\n",
       " 'start': None}"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_out.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# save the result\n",
    "with open('{}/{}_out_associations.json'.format(path, node), 'w') as f:\n",
    "    json.dump(r_out.json(), f, sort_keys=True, indent=4)\n",
    "with open('{}/{}_in_associations.json'.format(path, node), 'w') as f:\n",
    "    json.dump(r_in.json(), f, sort_keys=True, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evidence_graph</th>\n",
       "      <th>evidence_types</th>\n",
       "      <th>id</th>\n",
       "      <th>object</th>\n",
       "      <th>object_extension</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>publications</th>\n",
       "      <th>qualifiers</th>\n",
       "      <th>relation</th>\n",
       "      <th>slim</th>\n",
       "      <th>subject</th>\n",
       "      <th>subject_extension</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'nodes': [{'id': 'ECO:0000304', 'lbl': 'trace...</td>\n",
       "      <td>None</td>\n",
       "      <td>7b9b7ffd-9e85-494e-8f49-e5fbb7b7d508</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>[https://data.monarchinitiative.org/ttl/reacto...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{'nodes': [{'id': ':.well-known/genid/person-3...</td>\n",
       "      <td>None</td>\n",
       "      <td>af89e914-68e5-4b3f-8d02-ee6d214a8a19</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>[https://data.monarchinitiative.org/ttl/kegg.t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      evidence_graph evidence_types  \\\n",
       "0  {'nodes': [{'id': 'ECO:0000304', 'lbl': 'trace...           None   \n",
       "1  {'nodes': [{'id': ':.well-known/genid/person-3...           None   \n",
       "\n",
       "                                     id  \\\n",
       "0  7b9b7ffd-9e85-494e-8f49-e5fbb7b7d508   \n",
       "1  af89e914-68e5-4b3f-8d02-ee6d214a8a19   \n",
       "\n",
       "                                              object object_extension  \\\n",
       "0  {'deprecated': None, 'replaced_by': None, 'lab...             None   \n",
       "1  {'deprecated': None, 'replaced_by': None, 'lab...             None   \n",
       "\n",
       "                                         provided_by publications qualifiers  \\\n",
       "0  [https://data.monarchinitiative.org/ttl/reacto...         None       None   \n",
       "1  [https://data.monarchinitiative.org/ttl/kegg.t...         None       None   \n",
       "\n",
       "                                            relation  slim  \\\n",
       "0  {'deprecated': None, 'replaced_by': None, 'lab...  None   \n",
       "1  {'deprecated': None, 'replaced_by': None, 'lab...  None   \n",
       "\n",
       "                                             subject subject_extension  type  \n",
       "0  {'deprecated': None, 'replaced_by': None, 'lab...              None  None  \n",
       "1  {'deprecated': None, 'replaced_by': None, 'lab...              None  None  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_out_df = json_normalize(r_out.json(), 'associations')\n",
    "r_out_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evidence_graph</th>\n",
       "      <th>evidence_types</th>\n",
       "      <th>id</th>\n",
       "      <th>object</th>\n",
       "      <th>object_extension</th>\n",
       "      <th>provided_by</th>\n",
       "      <th>publications</th>\n",
       "      <th>qualifiers</th>\n",
       "      <th>relation</th>\n",
       "      <th>slim</th>\n",
       "      <th>subject</th>\n",
       "      <th>subject_extension</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'nodes': [{'id': ':.well-known/genid/610661.0...</td>\n",
       "      <td>None</td>\n",
       "      <td>005099b2-7ca9-4c41-a93d-20e02a43c9e7</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>[https://data.monarchinitiative.org/ttl/omim.t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{'nodes': [{'id': ':.well-known/genid/genoGM25...</td>\n",
       "      <td>None</td>\n",
       "      <td>e758883e-b6aa-46a1-a386-0e1f85539cad</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>[https://data.monarchinitiative.org/ttl/omim.t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'deprecated': None, 'replaced_by': None, 'lab...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      evidence_graph evidence_types  \\\n",
       "0  {'nodes': [{'id': ':.well-known/genid/610661.0...           None   \n",
       "1  {'nodes': [{'id': ':.well-known/genid/genoGM25...           None   \n",
       "\n",
       "                                     id  \\\n",
       "0  005099b2-7ca9-4c41-a93d-20e02a43c9e7   \n",
       "1  e758883e-b6aa-46a1-a386-0e1f85539cad   \n",
       "\n",
       "                                              object object_extension  \\\n",
       "0  {'deprecated': None, 'replaced_by': None, 'lab...             None   \n",
       "1  {'deprecated': None, 'replaced_by': None, 'lab...             None   \n",
       "\n",
       "                                         provided_by publications qualifiers  \\\n",
       "0  [https://data.monarchinitiative.org/ttl/omim.t...         None       None   \n",
       "1  [https://data.monarchinitiative.org/ttl/omim.t...         None       None   \n",
       "\n",
       "                                            relation  slim  \\\n",
       "0  {'deprecated': None, 'replaced_by': None, 'lab...  None   \n",
       "1  {'deprecated': None, 'replaced_by': None, 'lab...  None   \n",
       "\n",
       "                                             subject subject_extension  type  \n",
       "0  {'deprecated': None, 'replaced_by': None, 'lab...              None  None  \n",
       "1  {'deprecated': None, 'replaced_by': None, 'lab...              None  None  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_in_df = json_normalize(r_in.json(), 'associations')\n",
    "r_in_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>evidence_graph.edges</th>\n",
       "      <th>evidence_graph.nodes</th>\n",
       "      <th>evidence_types</th>\n",
       "      <th>id</th>\n",
       "      <th>object.categories</th>\n",
       "      <th>object.consider</th>\n",
       "      <th>object.deprecated</th>\n",
       "      <th>object.description</th>\n",
       "      <th>object.id</th>\n",
       "      <th>object.label</th>\n",
       "      <th>...</th>\n",
       "      <th>subject.id</th>\n",
       "      <th>subject.label</th>\n",
       "      <th>subject.replaced_by</th>\n",
       "      <th>subject.synonyms</th>\n",
       "      <th>subject.taxon.id</th>\n",
       "      <th>subject.taxon.label</th>\n",
       "      <th>subject.types</th>\n",
       "      <th>subject.xrefs</th>\n",
       "      <th>subject_extension</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[{'obj': ':.well-known/genid/610661.0004-61066...</td>\n",
       "      <td>[{'id': ':.well-known/genid/610661.0004-610661...</td>\n",
       "      <td>None</td>\n",
       "      <td>005099b2-7ca9-4c41-a93d-20e02a43c9e7</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>...</td>\n",
       "      <td>:.well-known/genid/person-GM25990</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[{'obj': ':.well-known/genid/genoGM25347', 'su...</td>\n",
       "      <td>[{'id': ':.well-known/genid/genoGM25347', 'lbl...</td>\n",
       "      <td>None</td>\n",
       "      <td>e758883e-b6aa-46a1-a386-0e1f85539cad</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>...</td>\n",
       "      <td>:.well-known/genid/person-3200-1</td>\n",
       "      <td>affected male proband with Congenital disorder...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[{'obj': 'OMIM:610661.0003', 'sub': ':.well-kn...</td>\n",
       "      <td>[{'id': ':.well-known/genid/610661.0003', 'lbl...</td>\n",
       "      <td>None</td>\n",
       "      <td>a9fa7673-df2a-4c2d-b4a5-0762fbef239d</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>...</td>\n",
       "      <td>:.well-known/genid/person-3213-1</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[{'obj': 'NCBIGene:55768', 'sub': 'OMIM:610661...</td>\n",
       "      <td>[{'id': 'Coriell:GM25340', 'lbl': 'NIGMS-GM253...</td>\n",
       "      <td>None</td>\n",
       "      <td>d26f2779-04c4-4614-836b-c88da38af258</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>...</td>\n",
       "      <td>:.well-known/genid/person-3199-1</td>\n",
       "      <td>affected male proband with Congenital disorder...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[{'obj': ':.well-known/genid/610661.0004-61066...</td>\n",
       "      <td>[{'id': ':.well-known/genid/610661.0004-610661...</td>\n",
       "      <td>None</td>\n",
       "      <td>1550dd15-3d25-40d7-883f-9f14fc2c554b</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>...</td>\n",
       "      <td>:.well-known/genid/person-GM25990</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                evidence_graph.edges  \\\n",
       "0  [{'obj': ':.well-known/genid/610661.0004-61066...   \n",
       "1  [{'obj': ':.well-known/genid/genoGM25347', 'su...   \n",
       "2  [{'obj': 'OMIM:610661.0003', 'sub': ':.well-kn...   \n",
       "3  [{'obj': 'NCBIGene:55768', 'sub': 'OMIM:610661...   \n",
       "4  [{'obj': ':.well-known/genid/610661.0004-61066...   \n",
       "\n",
       "                                evidence_graph.nodes evidence_types  \\\n",
       "0  [{'id': ':.well-known/genid/610661.0004-610661...           None   \n",
       "1  [{'id': ':.well-known/genid/genoGM25347', 'lbl...           None   \n",
       "2  [{'id': ':.well-known/genid/610661.0003', 'lbl...           None   \n",
       "3  [{'id': 'Coriell:GM25340', 'lbl': 'NIGMS-GM253...           None   \n",
       "4  [{'id': ':.well-known/genid/610661.0004-610661...           None   \n",
       "\n",
       "                                     id object.categories object.consider  \\\n",
       "0  005099b2-7ca9-4c41-a93d-20e02a43c9e7              None            None   \n",
       "1  e758883e-b6aa-46a1-a386-0e1f85539cad              None            None   \n",
       "2  a9fa7673-df2a-4c2d-b4a5-0762fbef239d              None            None   \n",
       "3  d26f2779-04c4-4614-836b-c88da38af258              None            None   \n",
       "4  1550dd15-3d25-40d7-883f-9f14fc2c554b              None            None   \n",
       "\n",
       "  object.deprecated object.description       object.id object.label  ...   \\\n",
       "0              None               None  NCBIGene:55768        NGLY1  ...    \n",
       "1              None               None  NCBIGene:55768        NGLY1  ...    \n",
       "2              None               None  NCBIGene:55768        NGLY1  ...    \n",
       "3              None               None  NCBIGene:55768        NGLY1  ...    \n",
       "4              None               None  NCBIGene:55768        NGLY1  ...    \n",
       "\n",
       "                          subject.id  \\\n",
       "0  :.well-known/genid/person-GM25990   \n",
       "1   :.well-known/genid/person-3200-1   \n",
       "2   :.well-known/genid/person-3213-1   \n",
       "3   :.well-known/genid/person-3199-1   \n",
       "4  :.well-known/genid/person-GM25990   \n",
       "\n",
       "                                       subject.label subject.replaced_by  \\\n",
       "0  affected female proband with Congenital disord...                None   \n",
       "1  affected male proband with Congenital disorder...                None   \n",
       "2  affected female proband with Congenital disord...                None   \n",
       "3  affected male proband with Congenital disorder...                None   \n",
       "4  affected female proband with Congenital disord...                None   \n",
       "\n",
       "  subject.synonyms subject.taxon.id subject.taxon.label subject.types  \\\n",
       "0             None   NCBITaxon:9606        Homo sapiens          None   \n",
       "1             None   NCBITaxon:9606        Homo sapiens          None   \n",
       "2             None   NCBITaxon:9606        Homo sapiens          None   \n",
       "3             None   NCBITaxon:9606        Homo sapiens          None   \n",
       "4             None   NCBITaxon:9606        Homo sapiens          None   \n",
       "\n",
       "  subject.xrefs subject_extension  type  \n",
       "0          None              None  None  \n",
       "1          None              None  None  \n",
       "2          None              None  None  \n",
       "3          None              None  None  \n",
       "4          None              None  None  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_in_df = json_normalize(r_in.json()['associations'])\n",
    "r_in_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "evidence_graph.edges    object\n",
       "evidence_graph.nodes    object\n",
       "evidence_types          object\n",
       "id                      object\n",
       "object.categories       object\n",
       "object.consider         object\n",
       "object.deprecated       object\n",
       "object.description      object\n",
       "object.id               object\n",
       "object.label            object\n",
       "object.replaced_by      object\n",
       "object.synonyms         object\n",
       "object.taxon.id         object\n",
       "object.taxon.label      object\n",
       "object.types            object\n",
       "object.xrefs            object\n",
       "object_extension        object\n",
       "provided_by             object\n",
       "publications            object\n",
       "qualifiers              object\n",
       "relation.categories     object\n",
       "relation.consider       object\n",
       "relation.deprecated     object\n",
       "relation.description    object\n",
       "relation.id             object\n",
       "relation.label          object\n",
       "relation.replaced_by    object\n",
       "relation.synonyms       object\n",
       "relation.types          object\n",
       "slim                    object\n",
       "subject.categories      object\n",
       "subject.consider        object\n",
       "subject.deprecated      object\n",
       "subject.description     object\n",
       "subject.id              object\n",
       "subject.label           object\n",
       "subject.replaced_by     object\n",
       "subject.synonyms        object\n",
       "subject.taxon.id        object\n",
       "subject.taxon.label     object\n",
       "subject.types           object\n",
       "subject.xrefs           object\n",
       "subject_extension       object\n",
       "type                    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_in_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject.id</th>\n",
       "      <th>subject.label</th>\n",
       "      <th>relation.id</th>\n",
       "      <th>relation.label</th>\n",
       "      <th>object.id</th>\n",
       "      <th>object.label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>:.well-known/genid/person-GM25990</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>:.well-known/genid/person-3200-1</td>\n",
       "      <td>affected male proband with Congenital disorder...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>:.well-known/genid/person-3213-1</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>:.well-known/genid/person-3199-1</td>\n",
       "      <td>affected male proband with Congenital disorder...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>:.well-known/genid/person-GM25990</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>:.well-known/genid/person-3199-1</td>\n",
       "      <td>affected male proband with Congenital disorder...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>:.well-known/genid/person-3213-1</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>:.well-known/genid/person-3200-1</td>\n",
       "      <td>affected male proband with Congenital disorder...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Coriell:GM25990</td>\n",
       "      <td>NIGMS-GM25990</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Coriell:GM25340</td>\n",
       "      <td>NIGMS-GM25340</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          subject.id  \\\n",
       "0  :.well-known/genid/person-GM25990   \n",
       "1   :.well-known/genid/person-3200-1   \n",
       "2   :.well-known/genid/person-3213-1   \n",
       "3   :.well-known/genid/person-3199-1   \n",
       "4  :.well-known/genid/person-GM25990   \n",
       "5   :.well-known/genid/person-3199-1   \n",
       "6   :.well-known/genid/person-3213-1   \n",
       "7   :.well-known/genid/person-3200-1   \n",
       "8                    Coriell:GM25990   \n",
       "9                    Coriell:GM25340   \n",
       "\n",
       "                                       subject.label relation.id  \\\n",
       "0  affected female proband with Congenital disord...        None   \n",
       "1  affected male proband with Congenital disorder...        None   \n",
       "2  affected female proband with Congenital disord...        None   \n",
       "3  affected male proband with Congenital disorder...        None   \n",
       "4  affected female proband with Congenital disord...        None   \n",
       "5  affected male proband with Congenital disorder...        None   \n",
       "6  affected female proband with Congenital disord...        None   \n",
       "7  affected male proband with Congenital disorder...        None   \n",
       "8                                      NIGMS-GM25990        None   \n",
       "9                                      NIGMS-GM25340        None   \n",
       "\n",
       "  relation.label       object.id object.label  \n",
       "0           None  NCBIGene:55768        NGLY1  \n",
       "1           None  NCBIGene:55768        NGLY1  \n",
       "2           None  NCBIGene:55768        NGLY1  \n",
       "3           None  NCBIGene:55768        NGLY1  \n",
       "4           None  NCBIGene:55768        NGLY1  \n",
       "5           None  NCBIGene:55768        NGLY1  \n",
       "6           None  NCBIGene:55768        NGLY1  \n",
       "7           None  NCBIGene:55768        NGLY1  \n",
       "8           None  NCBIGene:55768        NGLY1  \n",
       "9           None  NCBIGene:55768        NGLY1  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_df = r_in_df[['subject.id','subject.label', 'relation.id', 'relation.label', 'object.id', 'object.label']]\n",
    "in_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_queried_id</th>\n",
       "      <th>node_queried_label</th>\n",
       "      <th>relation_id</th>\n",
       "      <th>relation_label</th>\n",
       "      <th>node_associated_id</th>\n",
       "      <th>node_associated_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>RO:0002331</td>\n",
       "      <td>involved in</td>\n",
       "      <td>REACT:R-HSA-532668</td>\n",
       "      <td>N-glycan trimming in the ER and Calnexin/Calre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>OMIM:615273</td>\n",
       "      <td>NGLY1-deficiency</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_queried_id node_queried_label relation_id relation_label  \\\n",
       "0  NCBIGene:55768              NGLY1  RO:0002331    involved in   \n",
       "1  NCBIGene:55768              NGLY1        None           None   \n",
       "\n",
       "   node_associated_id                              node_associated_label  \n",
       "0  REACT:R-HSA-532668  N-glycan trimming in the ER and Calnexin/Calre...  \n",
       "1         OMIM:615273                                   NGLY1-deficiency  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save associations file\n",
    "r_out_df = json_normalize(r_out.json()['associations'])\n",
    "out_df = r_out_df[['subject.id','subject.label', 'relation.id', 'relation.label', 'object.id', 'object.label']]\n",
    "out_df = out_df.rename(\n",
    "                  columns={\n",
    "                      'subject.id':'node_queried_id',\n",
    "                      'subject.label':'node_queried_label',\n",
    "                      'relation.id':'relation_id',\n",
    "                      'relation.label':'relation_label',\n",
    "                      'object.id':'node_associated_id',\n",
    "                      'object.label':'node_associated_label'\n",
    "                      # extract also type/category and xref\n",
    "                  }\n",
    ")\n",
    "out_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_associated_id</th>\n",
       "      <th>node_associated_label</th>\n",
       "      <th>relation_id</th>\n",
       "      <th>relation_label</th>\n",
       "      <th>node_queried_id</th>\n",
       "      <th>node_queried_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>:.well-known/genid/person-GM25990</td>\n",
       "      <td>affected female proband with Congenital disord...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>:.well-known/genid/person-3200-1</td>\n",
       "      <td>affected male proband with Congenital disorder...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  node_associated_id  \\\n",
       "0  :.well-known/genid/person-GM25990   \n",
       "1   :.well-known/genid/person-3200-1   \n",
       "\n",
       "                               node_associated_label relation_id  \\\n",
       "0  affected female proband with Congenital disord...        None   \n",
       "1  affected male proband with Congenital disorder...        None   \n",
       "\n",
       "  relation_label node_queried_id node_queried_label  \n",
       "0           None  NCBIGene:55768              NGLY1  \n",
       "1           None  NCBIGene:55768              NGLY1  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_in_df = json_normalize(r_in.json()['associations'])\n",
    "in_df = r_in_df[['subject.id','subject.label', 'relation.id', 'relation.label', 'object.id', 'object.label']]\n",
    "in_df = in_df.rename(\n",
    "                columns={\n",
    "                      'subject.id':'node_associated_id',\n",
    "                      'subject.label':'node_associated_label',\n",
    "                      'relation.id':'relation_id',\n",
    "                      'relation.label':'relation_label',\n",
    "                      'object.id':'node_queried_id',\n",
    "                      'object.label':'node_queried_label'\n",
    "                      # extract also type/category and xref\n",
    "                }\n",
    ")\n",
    "in_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_queried_id</th>\n",
       "      <th>node_queried_label</th>\n",
       "      <th>relation_id</th>\n",
       "      <th>relation_label</th>\n",
       "      <th>node_associated_id</th>\n",
       "      <th>node_associated_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>RO:0002331</td>\n",
       "      <td>involved in</td>\n",
       "      <td>REACT:R-HSA-532668</td>\n",
       "      <td>N-glycan trimming in the ER and Calnexin/Calre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>OMIM:615273</td>\n",
       "      <td>NGLY1-deficiency</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_queried_id node_queried_label relation_id relation_label  \\\n",
       "0  NCBIGene:55768              NGLY1  RO:0002331    involved in   \n",
       "1  NCBIGene:55768              NGLY1        None           None   \n",
       "\n",
       "   node_associated_id                              node_associated_label  \n",
       "0  REACT:R-HSA-532668  N-glycan trimming in the ER and Calnexin/Calre...  \n",
       "1         OMIM:615273                                   NGLY1-deficiency  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "associations_df = pd.concat([out_df, in_df], ignore_index=True)\n",
    "#ssociations_df = associations_df[['node_queried_id', 'relation_id', 'node_associated_id']]\n",
    "associations_df = (\n",
    "    associations_df\n",
    "    [['node_queried_id', 'node_queried_label', 'relation_id', 'relation_label', 'node_associated_id', 'node_associated_label']]\n",
    ")\n",
    "associations_df.to_csv('{}/{}_associations.tsv'.format(path,node), sep='\\t', index=False, header=True)\n",
    "associations_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_queried_id</th>\n",
       "      <th>node_queried_label</th>\n",
       "      <th>relation_id</th>\n",
       "      <th>relation_label</th>\n",
       "      <th>node_associated_id</th>\n",
       "      <th>node_associated_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>RO:0002331</td>\n",
       "      <td>involved in</td>\n",
       "      <td>REACT:R-HSA-532668</td>\n",
       "      <td>N-glycan trimming in the ER and Calnexin/Calre...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>OMIM:615273</td>\n",
       "      <td>NGLY1-deficiency</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBIGene:55768</td>\n",
       "      <td>NGLY1</td>\n",
       "      <td>BFO:0000050</td>\n",
       "      <td>is part of</td>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>cytosol</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_queried_id node_queried_label  relation_id relation_label  \\\n",
       "0  NCBIGene:55768              NGLY1   RO:0002331    involved in   \n",
       "1  NCBIGene:55768              NGLY1         None           None   \n",
       "2  NCBIGene:55768              NGLY1  BFO:0000050     is part of   \n",
       "\n",
       "   node_associated_id                              node_associated_label  \n",
       "0  REACT:R-HSA-532668  N-glycan trimming in the ER and Calnexin/Calre...  \n",
       "1         OMIM:615273                                   NGLY1-deficiency  \n",
       "2          GO:0005829                                            cytosol  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#############################################\n",
    "# match id with network\n",
    "network_df = pd.DataFrame({'node_associated_id':['REACT:R-HSA-532668','OMIM:615273','GO:0005829']})\n",
    "connections_df = associations_df.merge(network_df, how='inner')\n",
    "connections_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# save to file\n",
    "connections_df.fillna('None').to_csv('{}/{}_associations_in_ngly1_graph.tsv'.format(path,node), sep='\\t', index=False, header=True)\n",
    "#############################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['REACT:R-HSA-532668',\n",
       " 'OMIM:615273',\n",
       " 'KEGG-path:map04141',\n",
       " 'GO:0005829',\n",
       " 'GO:0005515',\n",
       " 'GO:0006457',\n",
       " 'GO:0005737',\n",
       " 'GO:0046872',\n",
       " 'GO:0005634',\n",
       " 'GO:0006517',\n",
       " ':.well-known/genid/person-GM25990',\n",
       " ':.well-known/genid/person-3200-1',\n",
       " ':.well-known/genid/person-3213-1',\n",
       " ':.well-known/genid/person-3199-1',\n",
       " ':.well-known/genid/person-GM25990',\n",
       " ':.well-known/genid/person-3199-1',\n",
       " ':.well-known/genid/person-3213-1',\n",
       " ':.well-known/genid/person-3200-1',\n",
       " 'Coriell:GM25990',\n",
       " 'Coriell:GM25340']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get id -> qid\n",
    "# Wikidata 1:1 iff gene: 1:n.append(qid(protein))\n",
    "# list of monarch id\n",
    "associated_id_list = list(associations_df.node_associated_id)\n",
    "associated_id_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'results': {'bindings': [{'id': {'value': 'R-HSA-532668', 'type': 'literal'}, 'item': {'value': 'http://www.wikidata.org/entity/Q29017194', 'type': 'uri'}}]}, 'head': {'vars': ['id', 'item']}}\n",
      "R-HSA-532668 Q29017194\n"
     ]
    }
   ],
   "source": [
    "# wikidata query\n",
    "wd_type_property_dict = {'pathway':{'REACT':'P3937'}}\n",
    "node_type = 'pathway'\n",
    "node_id = 'REACT'\n",
    "node_code = '\"R-HSA-532668\"'\n",
    "\n",
    "query = \"\"\"SELECT DISTINCT ?id ?item\n",
    "WHERE\n",
    "{\n",
    "  ?item wdt:\"\"\" + wd_type_property_dict.get(node_type).get(node_id) + \"\"\" ?id .\n",
    "  values ?id {\"\"\" + node_code + \"\"\"}\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "# wikidata sparql api\n",
    "# api address:\n",
    "api = 'https://query.wikidata.org/sparql'\n",
    "\n",
    "# api call\n",
    "r = requests.post(api, data={'query': query}, headers={'Accept':'application/sparql-results+json'})\n",
    "\n",
    "# get results\n",
    "r.json()\n",
    "print(r.json())\n",
    "for wditem in r.json()['results']['bindings']:\n",
    "    id = wditem['id']['value']\n",
    "    ruri, qid = wditem['item']['value'].rsplit('/', 1)\n",
    "    print(id, qid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'results': {'bindings': [{'id': {'value': '55768', 'type': 'literal'}, 'item': {'value': 'http://www.wikidata.org/entity/Q21119040', 'type': 'uri'}}]}, 'head': {'vars': ['id', 'item']}}\n",
      "55768 Q21119040\n"
     ]
    }
   ],
   "source": [
    "# wikidata query for gene encoded proteins alternative\n",
    "wd_type_property_dict = {'gene':{'NCBIGene':'P351'}}\n",
    "node_type = 'gene'\n",
    "node_id = 'NCBIGene'\n",
    "node_code = '\"55768\"'\n",
    "\n",
    "query = \"\"\"SELECT DISTINCT ?id ?item \n",
    "WHERE\n",
    "{\n",
    "  ?geneItem wdt:\"\"\" + wd_type_property_dict.get(node_type).get(node_id) + \"\"\" ?id . # Entrez Gene ID\n",
    "  ?item wdt:P702 ?geneItem . # Encoded by\n",
    "  values ?id {\"\"\" + node_code + \"\"\"}\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "# wikidata sparql api\n",
    "# api address:\n",
    "api = 'https://query.wikidata.org/sparql'\n",
    "\n",
    "# api call\n",
    "r = requests.post(api, data={'query': query}, headers={'Accept':'application/sparql-results+json'})\n",
    "\n",
    "# get results\n",
    "r.json()\n",
    "print(r.json())\n",
    "for wditem in r.json()['results']['bindings']:\n",
    "    id = wditem['id']['value']\n",
    "    ruri, qid = wditem['item']['value'].rsplit('/', 1)\n",
    "    print(id, qid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'results': {'bindings': [{'id': {'value': 'http://www.wikidata.org/entity/Q18042037', 'type': 'uri'}, 'item': {'value': 'http://www.wikidata.org/entity/Q21119040', 'type': 'uri'}}]}, 'head': {'vars': ['id', 'item']}}\n",
      "http://www.wikidata.org/entity/Q18042037 Q21119040\n"
     ]
    }
   ],
   "source": [
    "# wikidata query for gene encoded proteins (the same query as above, \n",
    "# but gene has a list of properties that one is encodedBY)\n",
    "# here the input id is the qid for the gene\n",
    "wd_type_property_dict = {'gene':{'NCBIGene':'P351', 'encodedBy':'P702'}}\n",
    "node_type = 'gene'\n",
    "node_id = 'encodedBy'\n",
    "node_code = '<http://www.wikidata.org/entity/Q18042037>'\n",
    "#node_code = concat('<http://www.wikidata.org/entity/','Q18042037','>')\n",
    "\n",
    "query = \"\"\"SELECT DISTINCT ?id ?item \n",
    "WHERE\n",
    "{\n",
    "  ?item wdt:\"\"\" + wd_type_property_dict.get(node_type).get(node_id) + \"\"\" ?id .\n",
    "  values ?id {\"\"\" + node_code + \"\"\"}\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "# wikidata sparql api\n",
    "# api address:\n",
    "api = 'https://query.wikidata.org/sparql'\n",
    "\n",
    "# api call\n",
    "r = requests.post(api, data={'query': query}, headers={'Accept':'application/sparql-results+json'})\n",
    "\n",
    "# get results\n",
    "r.json()\n",
    "print(r.json())\n",
    "for wditem in r.json()['results']['bindings']:\n",
    "    id = wditem['id']['value']\n",
    "    ruri, qid = wditem['item']['value'].rsplit('/', 1)\n",
    "    print(id, qid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-40-857d438703fd>, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-40-857d438703fd>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    SELECT DISTINCT ?id ?item\u001b[0m\n\u001b[0m                  ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "    # are there id to nqid? yes, e.g. gene <http://www.wikidata.org/entity/Q18026920> => 35 proteins\n",
    "    #SELECT DISTINCT ?id (count(?item) as ?nitem)\n",
    "    #WHERE\n",
    "    #{\n",
    "    #  ?item wdt:P702 ?id .\n",
    "    #}\n",
    "    #group by ?id\n",
    "    #having (?nitem > 1)\n",
    "    #order by desc(?nitem)\n",
    "    #limit 10\n",
    "\n",
    "#SELECT DISTINCT ?id (group_concat(?item; separator='|') as ?items)\n",
    "SELECT DISTINCT ?id ?item\n",
    "    WHERE\n",
    "    {\n",
    "      ?item wdt:P702 ?id .\n",
    "      values ?id {<http://www.wikidata.org/entity/Q18026920>}\n",
    "    }\n",
    "#group by ?id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'8702': {'Q18033227': 1}, 'SO:0001587': {'Q28371788': 1}}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# function to map monarch_id to qid EASY: funcions una per les entitats i l'altra pels encoded proteins\n",
    "class Wikidata:\n",
    "    '''This class contains wikidata related data/methods'''\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.wdId2PropDict = {}\n",
    "    \n",
    "        self.wdId2PropDict['NCBIGene'] = 'P351' # Entrez Gene ID, code is geneid: 8702 *\n",
    "        self.wdId2PropDict['ZFIN'] = 'P3870' # ZFIN Gene ID, code is ZDB-GENE-030131-7764 *\n",
    "        self.wdId2PropDict['MGI'] = 'P671' # Mouse Genome Informatics ID, code is MGI:1913276 \n",
    "        self.wdId2PropDict['UniProt'] = 'P352' # UniProt ID, code is Q9UHM6 *\n",
    "        self.wdId2PropDict['UMLS'] = 'P2892' # UMLS CUI, code is C0023206 * # NO RESULTS cuz no ID statement\n",
    "        self.wdId2PropDict['HP'] = 'P3841' # Human Phenotype Ontology ID, code is HP:0002795\n",
    "        self.wdId2PropDict['FlyBase'] = 'P3852' # FlyBase Gene ID, code is FBgn0015872 *\n",
    "        self.wdId2PropDict['UBERON'] = 'P1554' # UBERON ID, code is 0000955 *\n",
    "        self.wdId2PropDict['DOID'] = 'P699' # Disease Ontology ID, code is DOID:0060068\n",
    "        self.wdId2PropDict['MESH'] = 'P486' # MeSH ID, code is D002277 *\n",
    "        self.wdId2PropDict['OMIM'] = 'P492' # OMIM ID, code is 608984 *\n",
    "        self.wdId2PropDict['Orphanet'] = 'P1550' # Orphanet ID, code is 404454 *\n",
    "        self.wdId2PropDict['NCBITaxon'] = 'P685' # Gene Ontology ID, code is 157 *\n",
    "        self.wdId2PropDict['REACT'] = 'P3937' # Reactome pathway ID, code is R-HSA-532668 *\n",
    "        self.wdId2PropDict['KEGG-path'] = 'P665' # KEGG ID, code is map00520 *\n",
    "        self.wdId2PropDict['CHEBI'] = 'P683' # ChEBI ID, code is 506227 *\n",
    "        self.wdId2PropDict['SO'] = 'P3986' # Sequence Ontology ID, code is SO:0001587   \n",
    "\n",
    "def getQid(dictionary = {}, node_id = 'NCBIGene', node_code = '\"8702\"'):\n",
    "    '''This functions returns id to qid.'''\n",
    "    \n",
    "    # variables\n",
    "    wdId2PropDict = Wikidata().wdId2PropDict\n",
    "    \n",
    "    # wd query\n",
    "    propertyQid = wdId2PropDict.get(node_id)\n",
    "    query = \"\"\"SELECT DISTINCT ?id ?item\n",
    "    WHERE\n",
    "    {\n",
    "      ?item wdt:\"\"\" + propertyQid + \"\"\" ?id .\n",
    "      values ?id {\"\"\" + node_code + \"\"\"}\n",
    "    }\n",
    "    \"\"\"\n",
    "    \n",
    "    # wikidata sparql api\n",
    "    # api address:\n",
    "    api = 'https://query.wikidata.org/sparql'\n",
    "\n",
    "    # api call\n",
    "    r = requests.post(api, data={'query': query}, headers={'Accept':'application/sparql-results+json'})\n",
    "    \n",
    "    # get results: 'idQidDict => {'id' : {'qid':1, 'qid':1}}\n",
    "    idQidDict = dictionary\n",
    "    if not idQidDict:\n",
    "        idQidDict = dict()\n",
    "    for wditem in r.json()['results']['bindings']:\n",
    "        id = wditem['id']['value']\n",
    "        ruri, qid = wditem['item']['value'].rsplit('/', 1)\n",
    "        idQidDict = utils.add_elem_with_dictionary(idQidDict, id, qid)\n",
    "        \n",
    "    \n",
    "    return idQidDict\n",
    "    \n",
    "#idQidDict = getQid(idQidDict)\n",
    "getQid(idQidDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# function to map monarch_id to qid EASY: funcions una per les entitats i l'altra pels encoded proteins\n",
    "def getQid2(node_type = 'm-protein', node_id = 'MESH', node_code = '\"D006023\"'):\n",
    "    '''This functions returns id to qid.'''\n",
    "    \n",
    "    # dictionaries {entityType:wd-properties}\n",
    "    # lack of: MP, CL, CHR,ClinVarVariant, FOODON\n",
    "    # mesh, umls, chebi,... can be in more than one type of node\n",
    "    wdId2PropDict = {}\n",
    "    \n",
    "    wdId2PropDict['gene'] = {'NCBIGene':'P351'} # Entrez Gene ID, code is geneid: 8702 *\n",
    "    wdId2PropDict['z-gene'] = {'ZFIN':'P3870'} # ZFIN Gene ID, code is ZDB-GENE-030131-7764 *\n",
    "    wdId2PropDict['m-gene'] = {'MGI':'P671'} # Mouse Genome Informatics ID, code is MGI:1913276 \n",
    "    wdId2PropDict['u-protein'] = {'UniProt':'P352'} # UniProt ID, code is Q9UHM6 *\n",
    "    wdId2PropDict['m-protein'] = {'MESH':'P486'} # MeSH ID, code is D006023 * # NO RESULTS cuz no ID statement\n",
    "    wdId2PropDict['um-protein'] = {'UMLS':'P2892'} # UMLS CUI, code is C0023206 * # NO RESULTS cuz no ID statement\n",
    "    wdId2PropDict['h-phenotype'] = {'HP':'P3841'} # Human Phenotype Ontology ID, code is HP:0002795\n",
    "    wdId2PropDict['f-phenotype'] = {'FlyBase':'P3852'} # FlyBase Gene ID, code is FBgn0015872 *\n",
    "    wdId2PropDict['anatomy'] = {'UBERON':'P1554'} # UBERON ID, code is 0000955 *\n",
    "    wdId2PropDict['d-disease'] = {'DOID':'P699'} # Disease Ontology ID, code is DOID:0060068\n",
    "    wdId2PropDict['m-disease'] = {'MESH':'P486'} # MeSH ID, code is D002277 *\n",
    "    wdId2PropDict['o-disease'] = {'OMIM':'P492'} # OMIM ID, code is 608984 *\n",
    "    wdId2PropDict['or-disease'] = {'Orphanet':'P1550'} # Orphanet ID, code is 404454 *\n",
    "    wdId2PropDict['taxon'] = {'NCBITaxon':'P685'} # Gene Ontology ID, code is 157 *\n",
    "    wdId2PropDict['r-pathway'] = {'REACT':'P3937'} # Reactome pathway ID, code is R-HSA-532668 *\n",
    "    wdId2PropDict['k-pathway'] = {'KEGG-path':'P665'} # KEGG ID, code is map00520 *\n",
    "    wdId2PropDict['chemical'] = {'CHEBI':'P683'} # ChEBI ID, code is 506227 *\n",
    "    wdId2PropDict['sequence'] = {'SO':'P3986'} # Sequence Ontology ID, code is SO:0001587\n",
    "    \n",
    "    # wd query\n",
    "    propertyQid = wdId2PropDict.get(node_type).get(node_id)\n",
    "    query = \"\"\"SELECT DISTINCT ?id ?item\n",
    "    WHERE\n",
    "    {\n",
    "      ?item wdt:\"\"\" + propertyQid + \"\"\" ?id .\n",
    "      values ?id {\"\"\" + node_code + \"\"\"}\n",
    "    }\n",
    "    \"\"\"\n",
    "    \n",
    "    # wikidata sparql api\n",
    "    # api address:\n",
    "    api = 'https://query.wikidata.org/sparql'\n",
    "\n",
    "    # api call\n",
    "    r = requests.post(api, data={'query': query}, headers={'Accept':'application/sparql-results+json'})\n",
    "    \n",
    "    # get results: 'idQidDict => {'id' : {'qid':1, 'qid':1}}\n",
    "    idQidDict = dict()\n",
    "    for wditem in r.json()['results']['bindings']:\n",
    "        id = wditem['id']['value']\n",
    "        ruri, qid = wditem['item']['value'].rsplit('/', 1)\n",
    "        idQidDict = utils.add_elem_with_dictionary(idQidDict, id, qid)\n",
    "        \n",
    "    \n",
    "    return idQidDict\n",
    "    \n",
    "getQid2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to map monarch_id to qid complicada: tot en una funcio\n",
    "def getQid(node_type = 'gene', node_id = 'NCBIGene', node_code = '\"8702\"'):\n",
    "    '''This functions returns id to qid.'''\n",
    "    \n",
    "    # dictionaries {entityType:wd-properties}\n",
    "    wd_type_property_dict = {}\n",
    "    wd_type_property_dict['pathway'] = {'REACT':'P3937'}\n",
    "    wd_type_property_dict['gene'] = {'NCBIGene':'P351', 'encodedBy':'P702'}\n",
    "    wd_type_property_dict['anatomy'] = {'UBERON':'P351'}\n",
    "    \n",
    "    \n",
    "    for key1 in wd_type_property_dict:\n",
    "        if 'NCBIGene' in wd_type_property_dict[key1]:\n",
    "            print('yes')\n",
    "        else:\n",
    "            print('no')\n",
    "    \n",
    "    # wd query\n",
    "    # if node_type is gene -> add uri_namespace to node_code\n",
    "    wd = '<http://www.wikidata.org/entity/'\n",
    "    if node_code.startswith('Q'):\n",
    "        node_code = pd.concat(wd,node_code,'>')\n",
    "    \n",
    "    query = \"\"\"SELECT DISTINCT ?id ?item\n",
    "    WHERE\n",
    "    {\n",
    "      ?item wdt:\"\"\" + wd_type_property_dict.get(node_type).get(node_id) + \"\"\" ?id .\n",
    "      values ?id {\"\"\" + node_code + \"\"\"}\n",
    "    }\n",
    "    \"\"\"\n",
    "    \n",
    "    # are there id to nqid? yes, e.g. gene <http://www.wikidata.org/entity/Q18026920> => 35 proteins\n",
    "    #SELECT DISTINCT ?id (count(?item) as ?nitem)\n",
    "    #WHERE\n",
    "    #{\n",
    "    #  ?item wdt:P702 ?id .\n",
    "    #}\n",
    "    #group by ?id\n",
    "    #having (?nitem > 1)\n",
    "    #order by desc(?nitem)\n",
    "    #limit 10\n",
    "    \n",
    "    \n",
    "    # wikidata sparql api\n",
    "    # api address:\n",
    "    api = 'https://query.wikidata.org/sparql'\n",
    "\n",
    "    # api call\n",
    "    r = requests.post(api, data={'query': query}, headers={'Accept':'application/sparql-results+json'})\n",
    "    \n",
    "    # get results\n",
    "    idQid_dct = {}\n",
    "    for wditem in r.json()['results']['bindings']:\n",
    "        id = wditem['id']['value']\n",
    "        ruri, qid = wditem['item']['value'].rsplit('/', 1)\n",
    "        #idQid[id] = qid\n",
    "        print(id, qid)\n",
    "        \n",
    "    \n",
    "    #return dict\n",
    "    \n",
    "getQid()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# get {'id' : list(qids)} dict from wikidata\n",
    "for assoc_id in associated_id_list:\n",
    "    namespace, nid = assoc_id.split(':')\n",
    "    if namespace == 'REACT':\n",
    "        # query wikidata - return a dict {id:list(qid)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function get monarch associations\n",
    "def getMonarchAssociationsResponse(node = 'NCBIGene:55768'):\n",
    "    '''This function performs api calls to Monarch for in and out edges. \n",
    "        It returns associated nodes'''\n",
    "    \n",
    "    # api addresses\n",
    "    biolink = 'https://api.monarchinitiative.org/api/association'\n",
    "    \n",
    "    # parameters\n",
    "    #data = {'use_compact_associations':'true','fl_excludes_evidence':'true'}\n",
    "    data = {'fl_excludes_evidence':'true'}\n",
    "    \n",
    "    # out edges: from/\n",
    "    r_out = requests.get('{}/from/{}'.format(biolink,node),data)\n",
    "\n",
    "    # in edges: to/\n",
    "    r_in = requests.get('{}/to/{}'.format(biolink,node),data)\n",
    "\n",
    "    return r_out, r_in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# prepare monarch api response as a dataframe or tuple?\n",
    "def prepareMonarchAssociationsResponseAsDf(r_out, r_in):\n",
    "    '''This function prepare the api object response from monarch to a dataframe. \n",
    "       It returns a dataframe.\n",
    "    '''\n",
    "    \n",
    "    # r_out response object to a dataframe\n",
    "    \n",
    "    r_out_df = json_normalize(r_out.json()['associations'])\n",
    "    out_df = r_out_df[['subject.id','subject.label', 'relation.id', 'relation.label', 'object.id', 'object.label']]\n",
    "    out_df = out_df.rename(\n",
    "                  columns={\n",
    "                      'subject.id':'node_queried_id',\n",
    "                      'subject.label':'node_queried_label',\n",
    "                      'relation.id':'relation_id',\n",
    "                      'relation.label':'relation_label',\n",
    "                      'object.id':'node_associated_id',\n",
    "                      'object.label':'node_associated_label'\n",
    "                      # extract also type/category and xref\n",
    "                  }\n",
    "    )\n",
    "    out_df.head(2)\n",
    "\n",
    "    # r_in response object to a dataframe\n",
    "    r_in_df = json_normalize(r_in.json()['associations'])\n",
    "    in_df = r_in_df[['subject.id','subject.label', 'relation.id', 'relation.label', 'object.id', 'object.label']]\n",
    "    in_df = in_df.rename(\n",
    "                columns={\n",
    "                      'subject.id':'node_associated_id',\n",
    "                      'subject.label':'node_associated_label',\n",
    "                      'relation.id':'relation_id',\n",
    "                      'relation.label':'relation_label',\n",
    "                      'object.id':'node_queried_id',\n",
    "                      'object.label':'node_queried_label'\n",
    "                      # extract also type/category and xref\n",
    "                }\n",
    "    )\n",
    "    in_df.head(2)\n",
    "    \n",
    "    # concat response dataframes\n",
    "    associations_df = pd.concat([out_df, in_df], ignore_index=True)\n",
    " \n",
    "    return associations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{('NCBIGene:55768', None, 'OMIM:615273'), ('NCBIGene:55768', None, 'Coriell:GM25990'), ('NCBIGene:55768', None, 'Coriell:GM25340'), ('NCBIGene:55768', 'RO:0002327', 'GO:0046872'), ('NCBIGene:55768', 'RO:0002327', 'GO:0005515'), ('NCBIGene:55768', None, ':.well-known/genid/person-3213-1'), ('NCBIGene:55768', 'BFO:0000050', 'GO:0005634'), ('NCBIGene:55768', 'RO:0002331', 'REACT:R-HSA-532668'), ('NCBIGene:55768', 'BFO:0000050', 'GO:0005829'), ('NCBIGene:55768', 'RO:0002331', 'KEGG-path:map04141'), ('NCBIGene:55768', 'BFO:0000050', 'GO:0005737'), ('NCBIGene:55768', None, ':.well-known/genid/person-3200-1'), ('NCBIGene:55768', None, ':.well-known/genid/person-3199-1'), ('NCBIGene:55768', 'RO:0002331', 'GO:0006517'), ('NCBIGene:55768', None, ':.well-known/genid/person-GM25990'), ('NCBIGene:55768', 'RO:0002331', 'GO:0006457')}\n",
      "{(None, None)}\n",
      "{(None, None, None)}\n",
      "{(None, None, None)}\n"
     ]
    }
   ],
   "source": [
    "# prepare monarch api response as a dataframe or tuple?\n",
    "def prepareMonarchAssociationsResponseAsTuple(r_out, r_in):\n",
    "    '''This function prepare the api object response from monarch to a tuple. \n",
    "       It returns a tuple of edges: {(s,r,o),(s,r,o),...}.\n",
    "    '''\n",
    "    \n",
    "    # variables\n",
    "    tripleId = set()\n",
    "    tripleXref = set()\n",
    "    tripleType = set()\n",
    "    tripleCategory = set()\n",
    "\n",
    "    # r_out response object to a dataframe\n",
    "    for association in r_out.json()['associations']:\n",
    "        tripleId.add((association['subject']['id'], association['relation']['id'], association['object']['id']))\n",
    "        tripleXref.add((association['subject']['xrefs'], association['object']['xrefs']))\n",
    "        tripleType.add((association['subject']['types'], association['relation']['types'], association['object']['types']))\n",
    "        tripleCategory.add((association['subject']['categories'], association['relation']['categories'], association['object']['categories']))\n",
    "\n",
    "    # r_in response object to a dataframe\n",
    "    for association in r_in.json()['associations']:\n",
    "        tripleId.add((association['object']['id'], association['relation']['id'], association['subject']['id']))\n",
    "        tripleXref.add((association['object']['xrefs'], association['subject']['xrefs']))\n",
    "        tripleType.add((association['object']['types'], association['relation']['types'], association['subject']['types']))\n",
    "        tripleCategory.add((association['object']['categories'], association['relation']['categories'], association['subject']['categories']))\n",
    "\n",
    "    # print    \n",
    "    print(tripleId)\n",
    "    print(tripleXref)\n",
    "    print(tripleType)\n",
    "    print(tripleCategory)\n",
    "    \n",
    "    #return tripleId\n",
    "\n",
    "prepareMonarchAssociationsResponseAsTuple(r_out, r_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{(None, None, None)}\n"
     ]
    }
   ],
   "source": [
    "# prepare monarch api response as a dataframe or tuple?\n",
    "def prepareMonarchAssociationsResponseAsTupleAlt(r_out, r_in):\n",
    "    '''This function prepare the api object response from monarch to a tuple. \n",
    "       It returns three lists, one for subjects, relations, and objects, \n",
    "       where each list is a list of dictionaries where each dict is a node.\n",
    "    '''\n",
    "    \n",
    "    # variables\n",
    "    sub_l = list()\n",
    "    rel_l = list()\n",
    "    obj_l = list()\n",
    "    tripleId = set()\n",
    "    tripleXref = set()\n",
    "    tripleType = set()\n",
    "    tripleCategory = set()\n",
    "\n",
    "    # r_out response object to a dataframe\n",
    "    # list of dictionaries where each dict is a node\n",
    "    for association in r_out.json()['associations']:\n",
    "        sub_l.append(association['subject'])\n",
    "        rel_l.append(association['relation'])\n",
    "        obj_l.append(association['object'])  \n",
    "\n",
    "    # r_in response object to a dataframe\n",
    "    for association in r_in.json()['associations']:\n",
    "        sub_l.append(association['object'])\n",
    "        rel_l.append(association['relation'])\n",
    "        obj_l.append(association['subject'])  \n",
    "    \n",
    "    return sub_l, rel_l, obj_l\n",
    "\n",
    "def edges(sub_l, rel_l, obj_l, attribute = 'id'):\n",
    "    '''\n",
    "       Function that builds triples with an attribute for each node. it returns a tuple with the edges\n",
    "    '''\n",
    "    edges = set()\n",
    "    # compose tuple\n",
    "    for i in range(len(sub_l)):\n",
    "        sub = sub_l[i][attribute]\n",
    "        rel = rel_l[i][attribute]\n",
    "        obj = obj_l[i][attribute]\n",
    "        edges.add((sub, rel, obj))\n",
    "        \n",
    "    return edges\n",
    "\n",
    "sub_l, rel_l, obj_l = prepareMonarchAssociationsResponseAsTupleAlt(r_out, r_in)\n",
    "edges = edges(sub_l, rel_l, obj_l, 'types')\n",
    "print(edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to map monarch_id to qid\n",
    "def getQid(associations_df):\n",
    "    '''This functions returns id to qid.'''\n",
    "    \n",
    "    # are there id to nqid?\n",
    "    \n",
    "    return associations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to evaluate the intersection with network nodes\n",
    "def filtering(nodes, edges):\n",
    "    \"\"\"filters all edges down to those which have both nodes in the nodes set\"\"\"\n",
    "    \n",
    "    keep = set()\n",
    "    for (start, pred, stop) in edges:\n",
    "        if {start, stop} <= nodes:\n",
    "            keep.add((start, pred, stop))\n",
    "            \n",
    "    return keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to evaluate the intersection with network nodes\n",
    "def getMatch(associations_df, network_df):\n",
    "    '''This function returns associated nodes that are in the network.'''\n",
    "    \n",
    "    # network and associations df have to be in qid \n",
    "    \n",
    "    #network_df = pd.DataFrame({'node_associated_id':['REACT:R-HSA-532668','OMIM:615273','GO:0005829']})\n",
    "    connections_df = associations_df.merge(network_df, how='inner')\n",
    "    return connections_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def printFile(path,fileName):\n",
    "    '''Function to save output to file.'''\n",
    "    connections_df.fillna('None').to_csv('{}/{}.tsv'.format(path,fileName), sep='\\t', index=False, header=True)\n",
    "    print(\"File './{}/{}' saved.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for node in nodes_l:\n",
    "    d = getMonarchAssociations()\n",
    "    e = getIntersection()\n",
    "    print(d + '_' + e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
